ml-agents-envs/mlagents_envs/rpc_utils.py

```diff
@@ -43,24 +46,57 @@ def behavior_spec_from_proto(
 
 
 @timed
-def process_pixels(image_bytes: bytes, gray_scale: bool) -> np.ndarray:
+def process_pixels(image_bytes: bytes, expected_channels: int) -> np.ndarray:
     """
     Converts byte array observation image into numpy array, re-sizes it,
     and optionally converts it to grey scale
-    :param gray_scale: Whether to convert the image to grayscale.
     :param image_bytes: input byte array corresponding to image
+    :param expected_channels: Expected output channels
     :return: processed numpy array of observation from environment
     """
-    with hierarchical_timer("image_decompress"):
-        image_bytearray = bytearray(image_bytes)
-        image = Image.open(io.BytesIO(image_bytearray))
-        # Normally Image loads lazily, this forces it to do loading in the timer scope.
-        image.load()
-    s = np.array(image, dtype=np.float32) / 255.0
-    if gray_scale:
+    image_bytearray = bytearray(image_bytes)
+
+    if expected_channels == 1:
+        # Convert to grayscale
+        with hierarchical_timer("image_decompress"):
+            image = Image.open(io.BytesIO(image_bytearray))
+            # Normally Image loads lazily, load() forces it to do loading in the timer scope.
+            image.load()
+        s = np.array(image, dtype=np.float32) / 255.0
         s = np.mean(s, axis=2)
         s = np.reshape(s, [s.shape[0], s.shape[1], 1])
-    return s
+        return s
+
+    image_arrays = []
+
+    bytes_read = 0
+    # Read the images back from the bytes (without knowing the sizes).
+    while bytes_read < len(image_bytearray):
+        # TODO avoid creating a new array here. Unfortunately, Pillow doesn't respect the current state of the buffer
+        # and always starts with seek(0), but we should be able to wrap BytesIO with something that lets us adjust
+        # the "start" offset.
+        buffer = io.BytesIO(image_bytearray[bytes_read:])
+        with hierarchical_timer("image_decompress"):
+            image = Image.open(buffer)
+            image.load()
+        image_arrays.append(np.array(image, dtype=np.float32) / 255.0)
+
+        # Look for the next header, starting from the current stream location
+        try:
+            offset = buffer.getvalue().index(PNG_HEADER, buffer.tell())
+            bytes_read += offset
```

Comment:
Given this code, I don't think we could ever hit the `while` condition.

Reply:
Good point, I'll change that to a `while True:`